examples/sst2/models.py

```diff
@@ -0,0 +1,385 @@
+
+  Example:
+  ```
+  sequence_mask([1, 2], 3)
+  [[True, False, False],
+   [True, True, False]]
+  ```
+
+  Args:
+    lengths: The length of each sequence. <int>[batch_size]
+    max_length: The width of the boolean mask. Must be >= max(lengths).
+
+  Returns:
+    A mask with shape: <bool>[batch_size, max_length] indicating which
+    positions are valid for each sequence.
+  """
+  return jnp.arange(max_length) < jnp.expand_dims(lengths, 1)
+
+
+@jax.vmap
+def flip_sequences(inputs: Array, lengths: Array) -> Array:
+  """Flips a sequence of inputs along the time dimension.
+
+  This function can be used to prepare inputs for the reverse direction of a
+  bidirectional LSTM. It solves the issue that, when naively flipping multiple
+  padded sequences stored in a matrix, the first elements would be padding
+  values for those sequences that were padded. This function keeps the padding
+  at the end, while flipping the rest of the elements.
+
+  Example:
+  ```python
+  inputs = [[1, 0, 0],
+            [2, 3, 0]
+            [4, 5, 6]]
+  lengths = [1, 2, 3]
+  flip_sequences(inputs, lengths) = [[1, 0, 0],
+                                     [3, 2, 0],
+                                     [6, 5, 4]]
+  ```
+
+  Args:
+    inputs: An array of input IDs <int>[batch_size, seq_length].
+    lengths: The length of each sequence <int>[batch_size].
+
+  Returns:
+    An ndarray with the flipped inputs.
+  """
+  max_length = inputs.shape[0]
+  return jnp.flip(jnp.roll(inputs, max_length - lengths, axis=0), axis=0)
+
+
+class WordDropout(nn.Module):
+  """Applies word dropout to a batch of input IDs.
+
+  This is basically the same as `nn.Dropout`, but allows specifying the
+  value of dropped out items.
+  """
+  dropout_rate: float
+  unk_idx: int
+  deterministic: bool = False
+
+  @nn.compact
+  def __call__(self, inputs: Array):
+    if self.deterministic or np.isclose(self.dropout_rate, 0.):
+      return inputs
+    rng = self.make_rng('dropout')
+    mask = jax.random.bernoulli(rng, p=self.dropout_rate, shape=inputs.shape)
+    return jnp.where(mask, jnp.array([self.unk_idx]), inputs)
+
+
+class Embedder(nn.Module):
+  """Embeds batches of token IDs into feature space.
+
+  Attributes:
+    vocab_size: The size of the vocabulary (i.e., the number of embeddings).
+    embedding_size: The dimensionality of the embeddings.
+    embedding_init: The initializer used to initialize the embeddings.
+    frozen: Freezes the embeddings table, keeping it fixed at initial values.
+    dropout_rate: Percentage of units to drop after embedding the inputs.
+    word_dropout_rate: Percentage of input words to replace with unk_idx.
+    unk_idx: The index (integer) to use to replace inputs for word dropout.
+  """
+  vocab_size: int
+  embedding_size: int
+  embedding_init: Callable[...,
+                           Array] = nn.initializers.normal(stddev=0.1)
+  frozen: bool = False
```

Comment:
Are you actually using this anywhere?

Reply:
I'm not using `frozen` in the current training.

Reply:
(But I might in the future :) I think it's nice to demonstrate how you could freeze the embedding if you would like to.)

Reply:
Ok, sounds good.